main.py

```diff
@@ -144,62 +145,145 @@ async def nini(ctx):
+    loss = {}
+
+    for name in all_losers:
+        if name not in old_losers:
+            loss[name] = all_losers[name]["errors"]
+        else:
+            loss[name] = all_losers[name]["errors"]-old_losers[name]["errors"]
+
+    loss_ranking = sorted([(name, nLoss) for name, nLoss in loss.items() if nLoss > 0], key=lambda x: x[1], reverse=True)
+
+    # RANKING UPDATE
+    old_losers = sorted([(name, data) for name, data in old_losers.items() if (data["messages"] >= 300 and data["errors"] > 0)],
                         key=lambda x: x[1]["messages"]/x[1]["errors"], reverse=True)
+
+    old_ranking = {}
+    for rank, name, data in enumerate(old_losers):
+        old_ranking[name] = rank
+
+    all_losers_sorted = sorted([(name, data) for name, data in all_losers.items() if (data["messages"] >= 300 and data["errors"] > 0)],
+                                key=lambda x: x[1]["messages"]/x[1]["errors"], reverse=True)
+
+    for new_rank, name, data in enumerate(all_losers_sorted):
+        if name not in old_ranking:
+            old_rank = len(old_ranking)
+        else:
+            old_rank = old_ranking[name]
+        update = old_rank-new_rank
+        if update > 0:
+            update = "🔼"+str(update)
+        elif update < 0:
+            update = "🔽"+str(-update)
+        else:
+            update = "🟦0"
+        data["rank_update"] = update
+
+    # RANKING MESSAGE
     res = ["Classement du <:nini:696420822855843910>"]
-    res.append("`**{}** | **{}** | **{}** | **{}** | **{}** | **{}**`".format("Pseudo".center(20), "# messages".center(12), "# défaites".center(12),
-                                                      "Ratio".center(7), "Streak".center(12), "Streak max".center(12)))
-    for username, data in all_losers:
-        res.append("`{} | {} | {} | {} | {} | {}`".format(username.center(20), str(data["messages"]).center(12), str(data["errors"]).center(12),
-                                                          str(int(data["messages"]/data["errors"])).center(7), str(data["streak"]).center(12), str(data["streak_max"]).center(12)))
+    res.append("`**{}** | **{}** | **{}** | **{}** | **{}** | **{}** | **{}**`".format("Pseudo".center(20), "# messages".center(12), "# défaites".center(12),
+                                                      "Ratio".center(7), "Streak".center(12), "Streak max".center(12), "Évolution".center(12)))
+
+    for username, data in all_losers_sorted:
+        res.append("`{} | {} | {} | {} | {} | {} | {}`".format(username.center(20), str(data["messages"]).center(12), str(data["errors"]).center(12),
+                                                          str(int(data["messages"]/data["errors"])).center(7), str(data["streak"]).center(12), str(data["streak_max"]).center(12), data["rank_update"].center(12)))
 
     log_info('\n'.join(res))
-    res.append("J'ai lu {} messages !".format(c))
+
+
+    # LOSER MESSAGE AND ROLES
+
+    nLoser = max(NUMBER_OF_LOSER, len(loss_ranking))
```

Comment:
If there is no loser, nLoser will be equal to NUMBER_OF_LOSER which is 2.

What's the purpose of this?

Reply:
Should be min instead of max

Reply:
Alsor resolved L206-208 which forgot to remove old user without new loss